wdl-grammar/src/experimental/parser.rs

```diff
@@ -0,0 +1,491 @@
+    where
+        T: ParserToken<'a>,
+    {
+        // Update the node kind and push a finished event
+        match &mut parser.events[self.0] {
+            Event::NodeStarted(k) => {
+                *k = kind;
+            }
+            _ => unreachable!(),
+        }
+
+        parser.events.push(Event::NodeFinished);
+        std::mem::forget(self);
+    }
+
+    /// Abandons the node due to an error.
+    pub fn abandon<'a, T>(self, parser: &mut Parser<'a, T>)
+    where
+        T: ParserToken<'a>,
+    {
+        // If the current node has no children, just pop it from the event list
+        if self.0 == parser.events.len() - 1 {
+            match parser.events.pop() {
+                Some(Event::NodeStarted(SyntaxKind::Abandoned)) => (),
+                _ => unreachable!(),
+            }
+        }
+
+        std::mem::forget(self);
+    }
+}
+
+impl Drop for Marker {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            panic!("marker was dropped without it being completed or abandoned");
+        }
+    }
+}
+
+/// Implements a WDL parser.
+///
+/// The parser produces a list of events that can be used to
+/// construct a CST.
+#[allow(missing_debug_implementations)]
+pub struct Parser<'a, T>
+where
+    T: ParserToken<'a>,
+{
+    /// The lexer that returns a stream of tokens for the parser.
+    lexer: Lexer<'a, T>,
+    /// The events produced by the parser.
+    events: Vec<Event>,
+}
+
+impl<'a, T> Parser<'a, T>
+where
+    T: ParserToken<'a>,
+{
+    /// Construct a new parser from the given lexer.
+    pub fn new(lexer: Lexer<'a, T>) -> Self {
+        Self {
+            lexer,
+            events: Default::default(),
+        }
+    }
+
+    /// Gets the current span of the parser.
+    pub fn span(&self) -> SourceSpan {
+        self.lexer.span()
+    }
+
+    /// Peeks at the next token from the lexer without consuming it.
+    ///
+    /// The token is not added to the event list.
+    pub fn peek(&mut self) -> Option<(T, SourceSpan)> {
+        while let Some((res, span)) = self.lexer.peek() {
+            if let Some(t) = self.consume_trivia(res, span, true) {
+                return Some(t);
+            }
+        }
+
+        None
+    }
+
+    /// Consumes the next token only if it matches the given token.
+    ///
+    /// Returns `true` if the token was consumed, `false` if otherwise.
+    pub fn next_if(&mut self, token: T) -> bool {
+        match self.peek() {
+            Some((t, _)) if t == token => {
+                self.next();
+                true
+            }
+            _ => false,
+        }
+    }
+
+    /// Parses a delimited list of nodes via a callback.
+    ///
+    /// The parsing stops when it encounters the `until` token or if
+    /// the callback returns `Ok(false)`.
+    pub fn delimited<F>(
+        &mut self,
+        delimiter: Option<T>,
+        until: TokenSet,
+        recovery: TokenSet,
+        mut cb: F,
+    ) -> Result<(), Error>
```

Comment:
This method doesn't need to return a `Result` as it's infallible (it'll internally handle recovery until it's done parsing the delimited items).